### PR TITLE
Remove references to coefficients from boundary conditions

### PR DESCRIPTION
**Pull request by cmacmackin**
This finally removes the `MFEMCoefficient` classes. That involves refactoring the boundary condition classes. There are now two versions of each type of boundary condition: one which takes constant values and one which takes a function name. It is a little messy needing to have these duplicate classes, but it is in-line with how MOOSE does these things.

### CODE REVIEW

**Comment by cmacmackin** [comment]
@alexanderianblair I'd like to add some more unit tests, to cover the essential BCs. However, it's not clear to me how we should do that. Reading the [GridFunction](https://docs.mfem.org/html/classmfem_1_1GridFunction.html) documentation, it doesn't look like it's trivial to check that a Dirichlet BC has been applied to the correct boundary. Do you have any suggestions?

**Comment by alexanderianblair** [comment]
Essential BCs remove (true) DoFs from the bilinear form (operator) corresponding to that variable in the `EquationSystem` - to check that the Dirichlet condition has been applied to the correct boundary, you could get the array of true DoFs from the `GridFunction`'s `ParFESpace` (via `fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list)` and check these DoFs on the trial `GridFunction` are set to the expected values from the BC. 

For a more robust test, you could also check that the essential DoFs have indeed been eliminated from the bilinear form of a test problem - may be overkill given existing tests in MFEM, though.

**Comment by cmacmackin** [comment]
@alexanderianblair I've written some initial tests for the Dirichlet boundary conditions (currently just checking that everything runs, not checking it actually does what's expected). The [tests for tangential vector Dirichlet boundaries](https://github.com/aurora-multiphysics/platypus/pull/53/files#diff-790a821f450f756b4cc8bc53e2300dce1ed72f316323988ec7a7f3843a262375R120) keep segfaulting and I'm at a bit of a loss as to why. I suspect it's because something hasn't been initialised correctly in the tests. Could you take a look and let me know if anything immediately stand out to you?

**Comment by alexanderianblair** [comment]
Looks like it's the FESpace being used - the tangential vector BC is expecting a FE type with a vector basis (ND), so is failing because it's trying to set scalar DoFs in the FESpace using a 3D vector at each DoF.

Using an FE variable based on an `H1_FECollection` for the nodal Dirichlet condition and a vectorial FE variable based on an `ND_FECollection` should fix this

**Comment by cmacmackin** [comment]
I think I've fixed the segfaults (on my computer at least). I'm not quite clear how to add the checks you described above, Would you be able to provide an example? That would probably be faster than trying to figure it out myself.

**Comment by lindsayad** [comment]
I personally like what you guys had before better. Just because MOOSE has different classes for constant, function, material property, postprocessor, etc. based coefficients doesn't mean you should change what you're doing. We've been actually moving towards more class consolidation in MOOSE by using the functor system, from which variables, functions, functor material properties, and postprocessors all derive so that we don't need quadruplicate classes to cover what a user might leverage in their input file

**Comment by lindsayad** [comment]
So there are currently three coefficient accessors on the problem (I think) that MFEM objects can use:

- `getProperties().getScalarProperty()`
- `getXXXFunctionCoefficient()`
- `makeXXXCoefficient()`

In practice all these objects care about is whether the coefficient is a `Coefficient`, `VectorCoefficient`, or `MatrixCoefficient`, so it would be neat if there was only one set of APIs for getting coefficients: `getScalarCoefficient`, `getVectorCoefficient`, `getMatrixCoefficient`. And then correspondingly you should have a smaller set of MFEM objects (like you did before) that do not care about whether the `(Vector/Matrix)Coefficient` was created through a function, material property, etc.

**Comment by lindsayad** [comment]
I personally enjoyed being able to use the same `Coefficient` object in multiple input file blocks in https://github.com/aurora-multiphysics/platypus/pull/62

**Comment by alexanderianblair** [comment]
> We've been actually moving towards more class consolidation in MOOSE by using the functor system, from which variables, functions, functor material properties, and postprocessors all derive so that we don't need quadruplicate classes to cover what a user might leverage in their input file

This is useful to know that this is the direction MOOSE is going in - we definitely could have `FunctionCoefficients` added to the `PropertiesManager`, allowing a consistent point of access for usage downstream (via `getProperties().getXXXProperty()`), which would keep the flexibility that you mention above for using coefficients in multiple blocks In that case, there are a few options for the design here:

1. We add `FunctionCoefficients` and `VectorFunctionCoefficients` to the associated `PropertiesManager`, so coefficients generated by materials, functions, or otherwise are all stored in the same place. This keeps MOOSE-like syntax for the setup of those coefficients, but a user wouldn't be able to name a function and a material the same name without issues, assuming the coefficient name is taken to be the same. Not necessarily the most intuitive for MOOSE users that a function or a material should be passed into downstream block as the coefficient input, but avoids the issues of duplication of downstream classes depending on their supported input.

2. We handle everything as `Coefficients` directly in the input - in which case we need to handle the dependencies between coefficients carefully when coefficients depend on each other (eg. for PWCoefficients used to represent materials). Setting up material properties in the Coefficients block like this would diverge from the familiar MOOSE method of setting up block restricted parameters in materials, which would be a barrier to entry. However, the name of the coefficients in the system is unambiguous this way.

IMO Option 1 would be more familiar for existing MOOSE users, and sounds closer to the UI MOOSE would have after moving to functors.

**Comment by lindsayad** [comment]
> Coefficients, on the other hand, have no information about the blocks they apply to.

This isn't true, is it? Don't you guys leverage `PWCoefficient::UpdateCoefficient` for particular mesh attributes/blocks? I guess I do not see any difference between block-restrictable material properties and `PWCoefficient`

**Comment by lindsayad** [comment]
Here is our `FunctorDirichletBC`:
```c++
InputParameters
FunctorDirichletBC::validParams()
{
  InputParameters params = ADDirichletBCBase::validParams();

  params.addClassDescription("Imposes the Dirichlet boundary condition "
                             "$u(t,\\vec{x})=h(t,\\vec{x})$, "
                             "where $h$ is a functor and can have complex dependencies.");

  params.addRequiredParam<MooseFunctorName>("functor", "The functor to impose");
  params.addParam<MooseFunctorName>(
      "coefficient", 1.0, "An optional functor coefficient to multiply the imposed functor");

  return params;
}

FunctorDirichletBC::FunctorDirichletBC(const InputParameters & parameters)
  : ADDirichletBCBaseTempl<Real>(parameters),
    _functor(getFunctor<ADReal>("functor")),
    _coef(getFunctor<ADReal>("coefficient"))
{
}

ADReal
FunctorDirichletBC::computeQpValue()
{
  const Moose::NodeArg space_arg = {_current_node, Moose::INVALID_BLOCK_ID};
  const Moose::StateArg time_arg = Moose::currentState();
  return _coef(space_arg, time_arg) * _functor(space_arg, time_arg);
}
```
Functors can be either variables, (functor) material properties, functions, or postprocessors. 

To me, `Coefficient` feels like a pretty good analog in MFEM to MOOSE's functor

**Comment by lindsayad** [comment]
> a user wouldn't be able to name a function and a material the same name without issues, assuming the coefficient name is taken to be the same.

We see this as a pro, not a con

**Comment by alexanderianblair** [comment]
> > Coefficients, on the other hand, have no information about the blocks they apply to.
> 
> This isn't true, is it? Don't you guys leverage `PWCoefficient::UpdateCoefficient` for particular mesh attributes/blocks? I guess I do not see any difference between block-restrictable material properties and `PWCoefficient`

Both of these statements are true - `mfem::Coefficient` doesn't have information about the blocks it applies to - this does indeed look like the analogue to MOOSE's Functors, from the above.  The derived class `mfem::PWCoefficient`, on the other hand (which is currently what's stored in the `PropertiesManager`) does have information on blocks - I agree with Alex that the analogue to an `mfem::PWCoefficient` (or a 'Property', currently) would be a block-restricted set of Material properties in MOOSE.


> I do not think it would be appropriate to mix the creation of Coefficients in with the PropertyManager class as it exists now. I suppose it would be possible to refactor it into some sort of coefficient manager class which takes on additional responsibilities, but I'm concerned that would lead to confusion about the distinction between coefficients and properties that I have described.

Right now, we store `PWCoefficients` (i.e. Materials) separately from `FunctionCoefficients` (i.e. Functions). When used, we only need to know that they're all derived from `Coefficients` (i.e. Functors). It's true that the `PropertyManager` was designed for setup of `PWCoefficients`, but it already supports the addition of 'global' coefficients if no block is given (via `PropertyMap::addProperty`). Addition of a `FunctionCoefficient` using `PropertyMap::addProperty` should work, and allow the `PropertyManager` to act as a single coefficient store - what would a `CoefficientManager` do differently to this? 

The `FunctorDirichletBC` snippet above looks identical to us supporting passing in a coefficient name in the input blocks - strongly supports moving to having a shared coefficient store IMO (Option 1 above). Where MOOSE would call `getFunctor`, we'd call `getCoefficient`, but from a user standpoint, the input files would be as similar as possible on this front.

**Comment by alexanderianblair** [comment]
After discussion with @cmacmackin:  we're conscious of the fact that Coefficients don't exist as a concept in vanilla MOOSE. As a result, we're keen to avoid them appearing as separate objects in HIT inputs as things currently stand, to minimise the barrier to entry for MOOSE users trying out the use of MFEM as an FE backend. That said, reducing the duplication of downstream objects (like BCs and Kernels) is very appealing.

Current plan of attack is:

- To remove Coefficients from HIT input files explicitly, by merging this PR as-is (following rebase)
- Open a new PR to investigate use of a unified Coefficients store in MFEMProblem, to avoid duplication of the input interface of BCs and Kernels for Functions and Materials. Possibly using `MFEMFunctorXXXBC` or `MFEMFunctorXXXKernel` classes, to keep the analogy with MOOSE.

**Comment by lindsayad** [comment]
The plan sounds reasonable to me

> * Open a new PR to investigate use of a unified Coefficients store in MFEMProblem, to avoid duplication of the input interface of BCs and Kernels for Functions and Materials. Possibly using `MFEMFunctorXXXBC` or `MFEMFunctorXXXKernel` classes, to keep the analogy with MOOSE.

Yea I like this. To elaborate more, I'd be fine with multiple database being unique/shared owners of different kinds of `Coefficients`. This is the case for MOOSE functors ... the variable warehouse owns variables, the function warehouse owns functions, etc. But then we have one container which doesn't have any ownership but holds pointers to all the different functors in the MOOSE simulation. This is the container that is queried by MooseObjects, using `getFunctor`. I imagine something similar could be done here

File: unit/src/MFEMEssentialBCTest.C [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+  mfem::ConstantCoefficient _scalar_zero;
+  mfem::VectorConstantCoefficient _vector_zero;
+
+  MFEMEssentialBCTest()
+    : MFEMObjectUnitTest("PlatypusApp"),
+      _scalar_fes(_mfem_mesh_ptr->getMFEMParMeshPtr().get(), 1, 3),
+      _vector_h1_fes(
+          _mfem_mesh_ptr->getMFEMParMeshPtr().get(), 1, 3, mfem::BasisType::GaussLobatto, 3),
+      _vector_hcurl_fes(_mfem_mesh_ptr->getMFEMParMeshPtr().get(), 2, 3, 3),
+      _vector_hdiv_fes(_mfem_mesh_ptr->getMFEMParMeshPtr().get(), 2, 3, 3),
```
Comment (alexanderianblair) [inline unit/src/MFEMEssentialBCTest.C:30]: This is currently causing the vector tests to fail for me - `vdim` should instead be equal to 1 for the vector FE types (scalar coefficients but vector basis functions, as opposed to the H1 vector case)

```suggestion
      _vector_hcurl_fes(_mfem_mesh_ptr->getMFEMParMeshPtr().get(), 2, 3),
      _vector_hdiv_fes(_mfem_mesh_ptr->getMFEMParMeshPtr().get(), 2, 3),
```

File: doc/content/source/bcs/MFEMScalarFunctionDirichletBC.C [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        !syntax parameters /
+        BCs /
+        MFEMScalarFunctionDirichletBC
+
+        !syntax inputs /
+        BCs /
+        MFEMScalarFunctionDirichletBC
+
+        !syntax children /
+        BCs / MFEMScalarFunctionDirichletBC
```
Comment (alexanderianblair) [inline doc/content/source/bcs/MFEMScalarFunctionDirichletBC.C:24]: Unexpected indent here 
```suggestion
fixing its values to the input scalar function on the boundary.

##Example Input File Syntax

!syntax parameters /
BCs /
MFEMScalarFunctionDirichletBC

!syntax inputs /
BCs /
MFEMScalarFunctionDirichletBC

!syntax children /
BCs / MFEMScalarFunctionDirichletBC
```

File: doc/content/source/bcs/MFEMVectorBoundaryIntegratedBC.md [hunk 3]
```diff
@@ -11,7 +11,7 @@ Adds the boundary integrator for integrating the linear form
 !equation
 (\vec f, \vec v)_{\partial\Omega} \,\,\, \forall \vec v \in V
 
-where $v \in \vec H^1$ and $\vec f$ is a vector coefficient of the same dimension.
+where $v \in \vec H^1$ and $\vec f$ is a constant vector of the same dimension.
```
Comment (alexanderianblair) [inline doc/content/source/bcs/MFEMVectorBoundaryIntegratedBC.md:14]:
```suggestion
where $\vec v \in \vec H^1$ and $\vec f$ is a constant vector of the same dimension.
```

File: unit/src/MFEMEssentialBCTest.C (hunk 1 above)
Comment (cmacmackin) [reply unit/src/MFEMEssentialBCTest.C:30]: Funny, given it worked on CI and on my computer.

**Comment by cmacmackin** [comment]
@alexanderianblair Could you run Clang-format on the code to get the linting working?
